dali/kernels/common/find/find_first_last_gpu.cuh

```diff
@@ -0,0 +1,194 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_COMMON_FIND_FIND_FIRST_LAST_GPU_CUH_
+#define DALI_KERNELS_COMMON_FIND_FIND_FIRST_LAST_GPU_CUH_
+
+#include <cuda_runtime.h>
+#include "dali/kernels/kernel.h"
+#include "dali/kernels/reduce/reductions.h"
+#include "dali/kernels/reduce/reduce_common.cuh"
+
+namespace dali {
+namespace kernels {
+namespace find {
+
+namespace {
+
+/**
+ * @brief Sample descriptor
+ */
+template <typename T, typename Predicate>
+struct SampleDesc {
+  int64_t *a_ptr, *b_ptr;  // represents (first, last), (begin, end), or (begin, length) depending
+                           // on the OutputProcessor
+  Predicate predicate;
+  const T *in;
+  int64_t len;
+};
+
+/**
+ * @brief Represents first, last (or other range representations) coordinates
+ */
+struct pair_i64 {
+    int64_t a = 0;
+    int64_t b = 0;
+};
+
+/**
+ * @brief First and last position
+ */
+struct first_last {
+  DALI_HOST_DEV DALI_FORCEINLINE pair_i64 operator()(pair_i64 x) const noexcept {
+    return x;
+  }
+
+  constexpr DALI_HOST_DEV DALI_FORCEINLINE pair_i64 neutral() const noexcept {
+    return {-1, -1};
+  }
+};
+
+/**
+ * @brief Begin (inclusive) and End (exclusive) of the region
+ */
+struct begin_end {
+  DALI_HOST_DEV DALI_FORCEINLINE pair_i64 operator()(pair_i64 x) const noexcept {
+    return {x.a, x.b + 1};
+  }
+
+  constexpr DALI_HOST_DEV DALI_FORCEINLINE pair_i64 neutral() const noexcept {
+    return {0, 0};  // empty range
+  }
+};
+
+/**
+ * @brief Begin (inclusive) and Length of the region
+ */
+struct begin_length {
+  DALI_HOST_DEV DALI_FORCEINLINE pair_i64 operator()(pair_i64 x) const noexcept {
+    return {x.a, x.b - x.a + 1};
+  }
+
+  constexpr DALI_HOST_DEV DALI_FORCEINLINE pair_i64 neutral() const noexcept {
+    return {0, 0};  // empty range
+  }
+};
+
+/**
+ * @brief Extract the position of the first and last position (or a derived representation) that
+ * satisfies a predicate
+ *
+ * @remarks Calculates a double reduction (min, max) in one go
+ *
+ * @tparam T Input data type
+ * @tparam Predicate Predicate that must be satisfied
+ * @tparam OutFormat Optional Transformation to first, last coordinates
+ * @param samples Sample descriptors, including a per-sample predicate
+ * @param format Optional output coordinate transformation
+ */
+template <typename T, typename Predicate, typename OutFormat = first_last>
+__global__ void FindFirstLastImpl(SampleDesc<T, Predicate> *samples, OutFormat format = {}) {
+  int sample_idx = blockIdx.y;
+  auto &sample = samples[sample_idx];
+  const T *input = sample.in;
+  int64_t sample_len = sample.len;
+  auto &predicate = sample.predicate;
+
+  const int64_t blk_size = blockDim.x * blockDim.y;
+  const int64_t grid_size = gridDim.x * blk_size;
```

Comment:
Why do you include `blockDim.y` in the math here? 
As I understand `blockDim.y` spans over different samples, so it should use only `blockDim.x` to iterate over a given sample.

Reply:
block is (32, 32). That's a requirement for the reduce kernels (what BlockReduce expects). That doesn't mean that the data is 2D. What the kernel typically does (both all the reduce kernels and this one) is to use a flattened block:
```
blk_size = blockDim.x * blockDim.y;  // 1024
threadIdx.x + threadIdx.y * blockDim.x   // 0...1023
```
So in short, the 2D layout of the threads is only used inside BlockReduce